src/Todo/App.purs

```diff
@@ -53,127 +53,188 @@ localStorageKey = "todomvc-purescript-state"
+          let
+            tasks' = Array.filter (not _.completed) self.state.tasks
+          in
+           Update self.state { tasks = tasks' }
+
+        CheckAllTasks targetChecked ->
+          let
+            toggle task = task { completed = fromMaybe task.completed targetChecked }
+            tasks' = map toggle self.state.tasks
+          in
+           Update self.state { tasks = tasks' }
+
+        UpdateVisibility vis ->
+          Update self.state { visibility = vis }
+
+        LoadState loadedState ->
+          Update loadedState
+
+        Noop ->
+          NoUpdate
 
 -- | Pure render function
-render :: forall r. { state :: State, setState :: SetState | r } -> JSX
-render { state, setState } =
+render :: Self Props State Action -> JSX
+render self =
   classy DOM.div "todomvc-wrapper"
     [ classy DOM.section "todoapp"
-      [ taskEntry state.newTodo onEditNewTodo onSubmitNewTodo
+      [ taskEntry self.state.newTodo onEditNewTodo onKbdNewTodo
       , taskList
-          { tasks: state.tasks
-          , visibility: state.visibility
+          { tasks: self.state.tasks
+          , visibility: self.state.visibility
           , onCheck: onTaskCheck
           , onDelete: onTaskDelete
           , onCommit: onTaskUpdate
           , checkAllTasks
           }
-      , React.element
-          Footer.component
-            { tasks: state.tasks
+      , Footer.component
+            { tasks: self.state.tasks
             , onClearCompleted: clearCompleted
-            , visibility: state.visibility
+            , visibility: self.state.visibility
             }
       ]
     ]
   where
     -- | Handler for editing the newTodo field
     onEditNewTodo =
-      Events.handler
-        (preventDefault >>> Events.merge { targetValue })
-        \{ targetValue } -> setState _ { newTodo = fromMaybe "" targetValue }
+      capture self targetValue EditNewTodo
 
     -- | Handler for submitting a new task after pressing enter
-    onSubmitNewTodo =
-      Events.handler
-        -- Events.merge lets us run two events in parallel
-        -- (in this case they are both pure)
-        -- and returns a record with their values
-        (Events.merge { targetValue, key })
-        \{ targetValue, key } -> case key of
-          Just "Enter" | not (String.null newDescription) -> do
-            _ <- pure preventDefault
-            setState _ { newTodo = ""
-                       , tasks = Array.cons newTodo state.tasks
-                       , uid = state.uid + 1
-                       }
-          _ -> pure unit
-            where
-              newDescription = String.trim state.newTodo
-
-              newTodo =
-                { description: newDescription
-                , id: state.uid
-                , completed: false
-                }
+    onKbdNewTodo =
+      monitor self key
+        \key ->
+          case key of
+            Just "Enter" -> do
+              KbdNewTodo
+            _ ->
+              Noop
```

Comment:
Style suggestion: I'd write this lambda as

```purescript
\case ->
  Just "Enter" -> KbdNewTodo
  _            -> Noop
```
(why don't we need to wrap the value `pure` in the catchall case?)

Reply:
Logic-wise: why not keep the guard for `String.null` here? In this way we have to check for `Noop` in two cases (here, and inside the `KbdNewTodo` action, where we send a `NoUpdate` in case the guard is not satisfied)

Reply:
It seems that guard statements cannot be used with LambdaCase. I decided to stay with the more verbose case statement and use the guard like the original version.